### PR TITLE
re-base of font fallback for pdf and eps output + SVG support

### PR DESCRIPTION
**Pull request by tacaswell**
## PR Summary

This is a rebase of #20804 and ##20832, given that there was not a lot of discussion on either PR, I am making a new PR rather than force-pushing to @aitikgupta 's branches.

My process here was to discard all changes to the c++ code and keep all changes to the Python code from #20804 and then implement the c++ layer needed to make the Python run (with some small changes to the Python + fixing up rebase errors.).
The current status of this:

 - the Agg tests still pass 🎉 
 - almost all the non-fallback tests pass 🎉 (I am seeing an svg bold font failure locally, not sure if it is related)
 - ~the pdf test only shows the Chinese characters 😞~ This was a "re use of names" bug  🎉 
 - ~c++ needs more error checking 😞~ This could still be better, but the one obvious place that needed it has it
 - recursive fallback is not implemented yet 😞  (but I think this is low-priority)
 - minimal test of `get_fontmap` 🎉 
 - ~the c++ style is terrible 😞~ acceptable now.

### CODE REVIEW

**Comment by tacaswell** [comment]
I have debugged this to something going wrong in writing the fonts and only one font being embedded in the PDF, but not why yet.

**Comment by tacaswell** [comment]
I have now squashed all of @aitikgupta 's work (+ adapting to the new c++ API) into two commits + a c++ commit + 2 extra commits.

I think this is ready to review.  The c++ needs the most attention.  There is likely also some un-needed work in the PDF backend.

**Comment by tacaswell** [comment]
Also looks like I accidentally used too-new of c++.

File: src/ft2font.cpp [hunk 1]
```diff
@@ -603,6 +603,34 @@ void FT2Font::load_char(long charcode, FT_Int32 flags, FT2Font *&ft_object, bool
     }
 }
 
+bool FT2Font::index_with_glyph(long charcode, int& index) const
```
Comment (QuLogic) [inline src/ft2font.cpp:607]: Not sure the name is clear. Also, do you need the `index`; can't it just return the `FT2Font`?

File: lib/matplotlib/backends/backend_pdf.py [hunk 2]
```diff
@@ -2358,7 +2367,10 @@ def draw_text(self, gc, x, y, s, prop, angle, ismath=False, mtext=None):
             fonttype = 1
         else:
             font = self._get_font_ttf(prop)
-            self.file._character_tracker.track(font, s)
+            char_to_font = font.get_fontmap(s)
+            self.file.fontName(prop)
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_pdf.py:2371]: Might be redundant?

File: lib/matplotlib/backends/backend_pdf.py [hunk 3]
```diff
@@ -2389,22 +2401,27 @@ def draw_text(self, gc, x, y, s, prop, angle, ismath=False, mtext=None):
...
                 if _font_supports_glyph(fonttype, ord(item.char)):
-                    if prev_was_multibyte:
-                        singlebyte_chunks.append((item.x, []))
+                    if prev_was_multibyte or item.ft_object != prev_font:
+                        singlebyte_chunks.append((item.ft_object, item.x, []))
+                        prev_font = item.ft_object
                     if item.prev_kern:
-                        singlebyte_chunks[-1][1].append(item.prev_kern)
-                    singlebyte_chunks[-1][1].append(item.char)
+                        singlebyte_chunks[-1][2].append(item.prev_kern)
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_pdf.py:2418]: Should this be only done if the font didn't change? Not sure you can mix kerning across fonts.

File: lib/matplotlib/backends/backend_ps.py [hunk 4]
```diff
@@ -647,21 +649,37 @@ def draw_text(self, gc, x, y, s, prop, angle, ismath=False, mtext=None):
                 thisx += kern * scale
                 xs_names.append((thisx, name))
                 thisx += width * scale
+            ps_name = (font.postscript_name
+                       .encode("ascii", "replace").decode("ascii"))
+            stream.append((ps_name, xs_names))
 
         else:
+            self.set_color(*gc.get_rgb())
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_ps.py:657]: This occurred on line 631 already?

File: lib/matplotlib/backends/backend_ps.py [hunk 5]
```diff
@@ -647,21 +649,37 @@ def draw_text(self, gc, x, y, s, prop, angle, ismath=False, mtext=None):
...
+                    curr_stream = [ps_name, []]
+                curr_stream[1].append(
+                    (item.x, item.ft_object.get_glyph_name(item.glyph_idx))
+                )
+            # append the last entry
+            stream.append(curr_stream)
+
+        for ps_name, xs_names in stream:
+            self.set_font(ps_name, prop.get_size_in_points(), False)
+            thetext = "\n".join(f"{x:f} 0 m /{name:s} glyphshow"
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_ps.py:680]: Why the change from `{x:g}` to `{x:f}`?

File: lib/matplotlib/backends/backend_pdf.py [hunk 6]
```diff
@@ -2358,7 +2367,9 @@ def draw_text(self, gc, x, y, s, prop, angle, ismath=False, mtext=None):
             fonttype = 1
         else:
             font = self._get_font_ttf(prop)
-            self.file._character_tracker.track(font, s)
+            char_to_font = font.get_fontmap(s)
+            for _c, _f in char_to_font.items():
+                self.file._character_tracker.track(_f, _c)
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_pdf.py:2372]: If this `for` loop happens everywhere, maybe it can go in `_character_tracker.track`?

File: src/ft2font.cpp (hunk 1 above)
Comment (QuLogic) [reply src/ft2font.cpp:607]: Also, should `long` be `FT_ULong`?

Comment (tacaswell) [reply src/ft2font.cpp:607]: This is searching in a list of c++ objects that do not know about the Python wrappers, we need to return index to the next layer up so it can look up in the list of PyObjects which are the wrappers around the corresponding c++ objects.

Doing it this way means that we do not need to have a cycle between the PyObjects and c++ objects.

File: lib/matplotlib/backends/backend_ps.py (hunk 5 above)
Comment (tacaswell) [reply lib/matplotlib/backends/backend_ps.py]: fixed.

File: src/ft2font.cpp (hunk 1 above)
Comment (QuLogic) [reply src/ft2font.cpp:607]: Yes, I thought `fallbacks` was the same thing here and in the caller, but I see that it's just the same name but for lists of different objects.

File: src/ft2font_wrapper.cpp [hunk 7]
```diff
@@ -552,6 +555,76 @@ static PyObject *PyFT2Font_get_kerning(PyFT2Font *self, PyObject *args)
...
+    PyObject *char_to_font;
+    if (!(char_to_font = PyDict_New())) {
+        return NULL;
+    }
+    for (auto it = codepoints.begin(); it != codepoints.end(); ++it) {
+        auto x = *it;
+        PyObject* target_font;
+        int index;
+        if (self->x->index_with_glyph(x, index)) {
+            if (index >=0) {
```
Comment (QuLogic) [inline src/ft2font_wrapper.cpp:604]: Space after `>=`

File: lib/matplotlib/backends/backend_pdf.py (hunk 3 above)
Comment (tacaswell) [reply lib/matplotlib/backends/backend_pdf.py:2415]: This is handled in `_text_helpers.py` (which is really handled at the c++ level) as the c++ layer knows to not try to do kerning between fonts.

**Review by QuLogic — APPROVED** [review]

**Review by QuLogic — APPROVED** [review]

File: lib/matplotlib/testing/_markers.py [hunk 8]
```diff
@@ -8,6 +8,7 @@
 import pytest
 
 import matplotlib.testing
+import matplotlib.testing.compare
```
Comment (oscargus) [inline lib/matplotlib/testing/_markers.py:11]: Is this required?

Comment (tacaswell) [reply lib/matplotlib/testing/_markers.py:11]: Yes, we use `matpotlib.testing.compare` in this module.  Under "normal" usage of running all of the tests, something else imports `matplotlib.testing.compare` before the code in this module gets run so it works.  However if you run just one module of tests (iirc, it was `test_backend_ps`) then that import does not happen an you get an `AttributeError`.

File: src/ft2font_wrapper.cpp [hunk 9]
```diff
@@ -552,6 +555,76 @@ static PyObject *PyFT2Font_get_kerning(PyFT2Font *self, PyObject *args)
...
+        for (size_t i = 0; i < size; ++i) {
+            codepoints.insert(unistr[i]);
+        }
+#else
+        for (size_t i = 0; i < size; ++i) {
+            codepoints.insert(PyUnicode_ReadChar(textobj, i));
+        }
+#endif
+    } else {
+        PyErr_SetString(PyExc_TypeError, "String must be str");
```
Comment (oscargus) [inline src/ft2font_wrapper.cpp:592]:
```suggestion
        PyErr_SetString(PyExc_TypeError, "string must be str");
```
No big deal, but in case you still push a new version...

File: lib/matplotlib/tests/test_ft2font.py [hunk 10]
```diff
@@ -76,3 +76,30 @@ def test_font_fallback_chinese(fig_test, fig_ref, family_name, file_name):
     ):
         fig_ref.text(0.05, .85 - 0.15*j, txt, family=ref_font)
         fig_test.text(0.05, .85 - 0.15*j, txt, family=test_font)
+
+
+@pytest.mark.parametrize(
+    "family_name, file_name",
+    [
+        ("WenQuanYi Zen Hei", "wqy-zenhei.ttc"),
+        ("Noto Sans CJK JP", "NotoSansCJK-Regular.ttc"),
```
Comment (oscargus) [inline lib/matplotlib/tests/test_ft2font.py:85]: Not sure I fully get this (probably not), but for OSX the font seems to be `NotoSansCJKsc-Regular.otf`. Not that it may be required to add that...

File: src/ft2font_wrapper.cpp (hunk 9 above)
Comment (tacaswell) [reply src/ft2font_wrapper.cpp]: I took this as matching the variable name seems better than matching grammer here.
